### ISSUE TITLE
ML: version learned model artifacts with provenance, metrics, and dataset manifests

### ISSUE THREAD

**Issue by cctsao1008** [cctsao1008/fami-pixel#34]
## Summary

The current SMB1 tiny surrogate correctly persists learned parameters (`w1`, `b1`, `w2`, `b2`) as a JSON artifact, currently defaulting to `build/models/smb1-tiny-risk.json`.

This is sufficient for inference, but it is not yet sufficient for long-term reproducibility, auditability, comparison between training runs, or reliable promotion of a learned model into planner experiments.

Introduce a versioned model-artifact contract so that a trained model is stored together with the provenance required to explain **where the weights came from, how they were trained, and how well they performed**.

## Motivation

A model file containing only weights and architecture tells us what the network is, but not why it should be trusted.

For every persisted surrogate we should be able to answer:

- Which rollout dataset produced these weights?
- Which repository commit and feature schema were used?
- Which train / validation / test split policy was used?
- Which hyperparameters and random seed were used?
- What were the validation, test, OOD, risk-recall, and ranking metrics?
- Is this artifact compatible with the current runtime feature extractor?
- Is this merely an experimental model, or an explicitly promoted baseline?

This is especially important because the learned surrogate is intentionally **not machine authority**. Its value is in ranking/pruning candidates before exact Mesen verification, so we need evidence that a given artifact improves search efficiency without damaging oracle coverage.

## Current state

The current model persistence contract in `src/fami_pixel/learning/tiny_mlp.py` stores:

- model format
- input size
- hidden size
- output definitions
- `w1`
- `b1`
- `w2`
- `b2`

The current training script writes the model by default to:

```text
build/models/smb1-tiny-risk.json
```

and separately prints training/evaluation information to stdout.

That means the model parameters survive process exit, but the full training provenance does not travel with the model artifact in a durable, structured form.

## Proposed artifact layout

Use a versioned per-model directory, for example:

```text
build/models/
└─ smb1-surrogate-v001/
   ├─ model.json
   ├─ metadata.json
   ├─ metrics.json
   └─ dataset_manifest.json
```

The exact path/naming may evolve, but the separation of concerns should remain explicit.

### `model.json`

Persist the inference artifact:

- format/version
- architecture
- outputs
- weights
- biases
- feature-schema identifier

### `metadata.json`

Record training provenance:

- model ID/version
- UTC creation timestamp
- git commit SHA
- dirty-tree flag when available
- training script / implementation version
- epochs
- learning rate
- hidden size
- random seed
- split strategy
- authority-boundary statement
- optional human note / experiment tag

### `metrics.json`

Persist evaluation results instead of only printing them:

- train metrics
- validation metrics
- test metrics
- chronological / OOD stress metrics
- delta-X MAE
- risk precision / recall
- no-progress precision / recall
- top-1 ranking accuracy
- top-2 / top-3 oracle coverage
- inference latency when measured

### `dataset_manifest.json`

Describe the actual training evidence without copying the dataset itself:

- input dataset paths or logical IDs
- file size / record count
- stable checksum (e.g. SHA-256) per dataset artifact
- root-state / generation grouping information where applicable
- split counts
- feature / target schema version
- optional collection-run identifiers

## Runtime compatibility

Add explicit validation so a model cannot be silently loaded against an incompatible feature extractor.

At minimum, loading should validate:

```text
model-format
feature-schema-id
input-size
output-schema
```

A mismatch should fail loudly rather than producing undefined inference behavior.

## Promotion concept

Do not make every generated model a source-controlled artifact.

Keep normal training outputs under `build/`, but define a small, explicit promotion mechanism for a model that becomes a reference baseline. Possible future shape:

```text
models/
└─ smb1-surrogate-baseline/
   ├─ model.json
   ├─ metadata.json
   ├─ metrics.json
   └─ dataset_manifest.json
```

Promotion should be deliberate and evidence-based, not automatic after every training run.

## Acceptance criteria

- [ ] Training produces a versioned model-artifact directory rather than only a standalone weights JSON.
- [ ] `model.json` preserves the current `TinySurrogateMLP` weights/biases and can be loaded for inference.
- [ ] `metadata.json` records code/training provenance and hyperparameters.
- [ ] `metrics.json` persists train/validation/test/OOD evaluation results.
- [ ] `dataset_manifest.json` records record counts and stable dataset fingerprints/checksums.
- [ ] A feature-schema/version identifier is introduced and stored with the model.
- [ ] Model loading validates schema compatibility and fails clearly on mismatch.
- [ ] Existing model loading remains backward-compatible where practical, or a migration path is documented.
- [ ] Tests cover artifact save/load round-trip and incompatible-schema rejection.
- [ ] Documentation explains generated vs promoted model artifacts.

## Non-goals

This issue does **not** grant the learned surrogate additional authority.

The existing control contract remains:

```text
learned model -> rank/prune candidates
exact Mesen rollout -> authoritative trajectory evidence
live Mesen -> machine authority
```

This issue is about making learned knowledge reproducible and auditable, not replacing Mesen with ML.

## Follow-up opportunities

After the artifact contract is stable, later work can evaluate:

- automated model comparison / regression reports;
- top-K oracle-coverage gates for model promotion;
- deterministic re-training checks;
- artifact retention policy;
- optional resume-training checkpoints separate from inference artifacts;
- ONNX or other runtime formats if the tiny dependency-free JSON model becomes insufficient.

**Comment by cctsao1008** [comment]
## Implementation checkpoint — versioned surrogate artifact contract is on `main`

Implemented the #34 core contract without changing ML authority semantics.

### New artifact contract

Added `src/fami_pixel/learning/model_artifact.py`.

Normal training now produces an auto-versioned directory under `build/models/`:

```text
smb1-surrogate-vNNN/
├─ model.json
├─ metadata.json
├─ metrics.json
└─ dataset_manifest.json
```

Key behavior:

- auto-increments `v001`, `v002`, ... without overwriting an existing artifact;
- `model.json` remains the runtime inference file and preserves `w1/b1/w2/b2`;
- `metadata.json` records UTC creation time, git SHA + dirty flag when available, trainer version, hyperparameters, split strategy, experiment tags/note, and the ML/Mesen authority boundary;
- `metrics.json` persists train/validation/test metrics, grouped chronological OOD stress metrics, training time, and measured inference latency;
- `dataset_manifest.json` records per-input path, byte size, SHA-256, record count, root-group count, source IDs, rollout/feature/target schema IDs, and split partition summaries.

### Explicit runtime schema compatibility

`TinySurrogateMLP` now writes model format `fami-pixel-tiny-surrogate-v2` with explicit:

- `feature_schema_id = smb1-tiny-surrogate-features-v1`
- fixed input width = 34
- output schema = `delta_x`, `risk_probability`, `no_progress_probability`

Loading fails loudly on model-format, feature-schema, input-size, or output-schema mismatch.

Backward compatibility is retained for the old `fami-pixel-tiny-surrogate-v1` standalone JSON when its implicit 34-input feature contract and output schema match the current extractor.

### Trainer integration

`tools/train_smb1_surrogate_mlp.py` now writes the four-file versioned artifact by default. `--output-artifact ...-vNNN` allows an explicit versioned directory. `--output-model` remains as an optional extra standalone V2 copy for old local workflows.

### Tests

Added:

- `tests/test_model_artifact.py`
  - four-file artifact round-trip
  - provenance/fingerprint checks
  - feature-schema rejection
  - input-size rejection
  - output-schema rejection
  - V1 backward-compatible load
  - version-number advancement
- `tests/test_train_surrogate_artifact.py`
  - one-epoch synthetic training smoke proving the trainer emits the expected four-file `smb1-surrogate-v001` artifact

Existing `tests/test_tiny_mlp.py` continues to cover prediction-preserving JSON round-trip and training metrics.

### Documentation

Updated `docs/research/smb1-surrogate.md` with:

- generated artifact layout;
- schema/migration contract;
- generated vs deliberately promoted model distinction;
- runtime `--surrogate-model .../model.json` usage;
- promotion/authority boundary.

### Commits in this slice

- `a0572060` — add versioned artifact helper
- `000555d3` — version model/feature/output schema
- `26626231` — trainer emits versioned artifacts
- `cd2de2d1`, `de187b0f` — artifact/schema tests
- `ceb293e5` — export artifact helpers
- `d7b9a027` — artifact/promotion documentation
- `ce5201fb` — stronger cross-schema validation
- `13e6a9e4` — trainer integration smoke

Authority boundary is unchanged:

```text
learned model -> rank/prune
exact Mesen rollout -> authoritative trajectory evidence
live Mesen -> machine authority
```

Final closure gate: run the focused local pytest slice (no Mesen/ROM required):

```powershell
py -m pytest tests/test_tiny_mlp.py tests/test_model_artifact.py tests/test_train_surrogate_artifact.py -q
```

Keep #34 open until that focused suite passes on the local checkout.

**Comment by cctsao1008** [comment]
Closure audit — focused acceptance PASS on Windows local repo.

Command:

```powershell
py -m pytest tests/test_tiny_mlp.py tests/test_model_artifact.py tests/test_train_surrogate_artifact.py -q
```

Result:

```text
........... [100%]
11 passed in 1.65s
```

Acceptance mapping:

- versioned artifact directory: implemented (`smb1-surrogate-vNNN/`)
- `model.json`: preserves TinySurrogateMLP weights/biases and loads for inference
- `metadata.json`: records Git/training provenance, hyperparameters, split strategy, authority boundary, tags/notes
- `metrics.json`: persists train/validation/test plus chronological OOD metrics and inference latency
- `dataset_manifest.json`: persists paths, sizes, SHA-256, record/root-group counts, schema IDs and split summaries
- explicit feature schema ID + fixed-width 34-input contract: implemented
- schema/output/input compatibility validation: implemented; mismatches fail loudly
- legacy V1 standalone model JSON loading: retained where shape/output contract is compatible
- tests cover artifact round-trip, checksum provenance, incompatible-schema rejection, legacy loading, version increment, and trainer end-to-end artifact emission
- documentation updated for generated vs promoted artifacts and legacy migration

Authority boundary is unchanged: learned model remains advisory rank/prune guidance; exact Mesen rollouts and live Mesen remain authoritative.

All #34 acceptance criteria are satisfied. Closing as completed.
